Add new filters Stage 2

McGiverGim opened the pull request:
This code adds the possibility to configure the new filters (Stage 2) of Betaflight.

This is a work in progress, because I have some doubts:
- The gyro lpf stage 1 is being only sent as U8 in the firmware and must be sended as U16 because it now is bigger (limit 16000).
- I think is a good idea to give a more "detailed" explanation in the tooltip of the filters. But I don't know exactly what to put there.
- Is a little strange than the Notch filters have a switch and all the others not. I think is better to add this to all filters or remove them from all. Thoughts?
- Is some other filter parameter interesting to be here and not only CLI? I'm talking about the type of gyro LPF, hardware filters, etc.

![image](https://user-images.githubusercontent.com/2673520/40916404-e0d4c6b4-67ff-11e8-84ef-0f1f201e4de6.png)

DieHertz commented:
It's no longer called `stage2` so it's better to omit the word `stage` altogether.
Call it lowpass1 and lowpass2, first and second, I don't know.

McGiverGim commented:
Ok, I will need to change it then in the blackbox too. I was calling them stage 2 too...

DieHertz commented:
Yep, this name was misleading and giving them a property that isn't really there, as they're order-invariant.

McGiverGim commented:
I've played a little with this, I don't know if this changes are good or do you prefer the old style:
![image](https://user-images.githubusercontent.com/2673520/40964287-043e8f9e-68ab-11e8-9ef7-e3e402fda6b1.png)

- I have moved the type of DTerm filter to the table, and not alone in the top
- Header for each "group" of filters
- Switch in the same line than the filter

So, to finish this PR, thing to decide:
- Old style or the new style
- Add or not more switchs to other kind of filters (activate/deactive), not only the notch.
- Add/remove elements (remove the type of dterm filter, add type of gyro filters, add hardware filters, etc.)

fujin commented:
@McGiverGim I like:
- New style - lots of dead space in the left td though.
- Switches that set the value to 0 (optional)
- d-term lowpass type must remain it can still be changed (only d-term lowpass2 is hard-coded to pt1)
- Gyro lowpass type would be nice to change, but with reduced options now I don't see it as crucial
- Gyro DLPF type/mode might be nice here, but should be considered expert functionality. The defaults for 8k and 32k mode are good enough for the everyday user I think. Power user can reach into CLI. I defer to you on that one. check with @etracer65 as well.

It'd be *really* cool to somehow do Bode plots or similar for all enabled filters and settings, overlay/combine them, as well as some statistics around group delay, similar to Mark Spatz' tool, but I do not expect this will come easily or urgently.

McGiverGim commented:
The only problem of adding more options is that it occupies more space and the scroll will be needed.
I will try to add some of them tomorrow.

fujin commented:
@McGiverGim perhaps another option could be to add another column, splitting the current width in half (allowing for the rows to be side-by-side). that would prevent the vertical scrolling. I agree it's worth avoiding

McGiverGim commented:
I thought it but I don't know if they will be more or less the same size. I will try it tomorrow.

etracer65 commented:
I wouldn't recommend adding the gyro DLPF settings (`gyro_hardware_lpf` & `gyro_32khz_hardware_lpf`).  Those are definitely expert user parameters and should only be in the CLI.

McGiverGim commented:
I'm working on it, the latest version will have:
- The new style
- I will try to adjust the first td space
- A switch next to each filter to deactivate it
- Select with types for gyro 1 and 2 and dterm 1.
- It will not have hardware filter options
- I will try to put it in two columns that change to one column when there is not enough space (lower resolution monitors or similar). I can't promise it but I will se if it works as expected.

A lot of changes. Some things that I have found while testing it:
- The dterm lpf 1 default is not the same in this version or in older (biquad 90 vs pt1 100). I don't know if it is a good idea to maintain different defaults here, and complicate more the code attached to versions, or simply have one default. This default is used when reactivating with a switch a filter.
- The yaw lpf by default is zero. So having a switch is strange, but I added it for coherence.
- Is a good idea let to put zero when the filter is activated? Maybe is better to put a minimum value (1 for example?)

mikeller commented:
@McGiverGim:

> The dterm lpf 1 default is not the same in this version or in older (biquad 90 vs pt1 100). I don't know if it is a good idea to maintain different defaults here, and complicate more the code attached to versions, or simply have one default. This default is used when reactivating with a switch a filter.

One default is completely fine. The aim here is **not** to replicate the firmware default, but to provide a value that is a good starting point.

McGiverGim commented:
More or less ready. Please test it to see if it can be better.

Finally it fits two columns with maximum and minimum size:

![image](https://user-images.githubusercontent.com/2673520/41045063-dcecbc6a-69a7-11e8-866b-41f6da7f3a8b.png)

![image](https://user-images.githubusercontent.com/2673520/41045076-e63ed1e0-69a7-11e8-96c6-d88bd6f65864.png)

I think I need to put a "default" to the yaw lowpass filter, but I don't know what value can be good for yaw, I never used.

McGiverGim commented:
Finally, I have:
- Limited to 1 the Hz of notch, 
- Default value for yaw 100
- Reordered and renamed all the filters to be coherent

It is not perfect, we can check more things, but I think it can be merged as is now.

mikeller commented on src/js/tabs/pid_tuning.js:
```diff
@@ -720,6 +793,29 @@ TABS.pid_tuning.initialize = function (callback) {
 
+
+        // DTerm filter options
+        function loadFilterTypeValues(){
```
Missing space.

McGiverGim commented:
All of them can be "completed" with more information but I don't know if they are interesting or not. Maybe explain than a lpf filter frecuencies starting at the value configured. 
If not, maybe remove all of them except the notch explanation?

mikeller commented:
@McGiverGim: If some of them need additional information we can re-add them when the information is added. The way they are now this should not be released, as it effectively 'hides' the tooltips with useful information, since users will not go and peek at all of them to find the 2 or 3 that have useful information.

McGiverGim commented:
Yes, that is the reason why I asked. Do I remove all except the notch explanation? If yes, let the same notch info in all three notch filters?

mikeller commented:
Yes please.

McGiverGim commented:
Changes done. Finally, after playing a bit with it, I have found a good place:
![image](https://user-images.githubusercontent.com/2673520/41106571-91bee316-6a70-11e8-8823-f92cb3deda63.png)

Now we have filters "groups", so I have added the help to the notch groups, in this way it has more sense. I've removed all the other help icons.

Please take a look to the help text, my english is not very good so maybe is a better way to explain a notch filter.

AndersHoglund commented:
Nice! 
Some consistent wording would be good though. You say "Frequency" and "Cutoff Frequency" in the menus for the notch filters but only "Frequency" and "Cutoff" in the help text. For clarity and consistency I think it should be spelled out as "Center Frequency" and "Cutoff Frequency" in both places.
"Frequency" is also used for the Lowpass filters, but here it actually mean "Cutoff Frequency". I think it should say so too in the menus.

McGiverGim commented:
Perfect @AndersHoglund !! 

Changes done to add consistency to the names:
![image](https://user-images.githubusercontent.com/2673520/41117018-2d95aeee-6a8c-11e8-9fac-0ebbc174d541.png)

mikeller commented:
Thanks for this, @McGiverGim!
